### DISCUSSION TITLE
Mapsui is bypassing my proxy

### DISCUSSION THREAD

**Discussion by AndersEkmanLm** [Mapsui/Mapsui#1519] [Q&A]
I'm using Mapsui for a desktop applicaiton, but I am having trouble making it use my WMS server (lets call it `maps.example.com`). The WMS server requires authentication via the authorization header. This authentication is provided by a reverse HTTP proxy (lets call it `proxy.example.com`) that sets the authroization header.

So I instruct Mapsui to use `proxy.example.com` without any authorization header. The proxy server than connects to `maps.example.com` with the authorization header set. This works fine for the GetCapabilities request, but after that all requests are made directly to `maps.example.com` instead that just returns a 401 since they are not authenticated.

The below example is an attempt to explain the situation: 

	var provider = new WmsProvider("proxy.example.com", "1.3.0")
	{
	  ContinueOnError = false,
	  TimeOut = 20000,
	  CRS = "EPSG:3857"
	};
	provider.AddLayer("my_layer");
	provider.SetImageFormat(provider.OutputFormats[0]);
	var rect = new Mapsui.Geometries.BoundingBox(6500000, 2000000, 6500000 + 100, 2000000 + 100);
	// This gives a URL to "maps.example.com", not "proxy.example.com"!
	var test = provider.GetRequestUrl(rect, 100, 100);
	
My guess is that Mapsui is reading the `maps.example.com` URL from the GetCapabilities response (it is included in several places there), and use that URL instead of the URL I provided. Is there any way to force Mapsui to always use the URL i provide, so that I can use my proxy?

**Comment by pauldendulk** [comment]
Two things that could help you.
- If you want to investigate what is happening here you could copy WmsProvider.cs into your own project (and perhaps rename to the class to MyWmsProvider) and debug it in there.
- Note that the WmsProvider constructor has a 'getStreamAsync' argument which allows you to override the web request, so this should not be blocking.

The proxy issue comes up in other situations as well (tiling) and this is old code so there is undoubtedly room for improvement. Perhaps we should be using IHttpClientFactoring. If you have suggestions to improve this then please let us know.

**Reply by AndersEkmanLm** [reply]
Thanks! I think your second bullet point look promising. So the solution would be to write my own download method, that calls the default download method but switches the URL back to `proxy` instead of `maps`. I will work on this later and report back here.

I don't think I am knowledgable enough to come with any good suggestions. I don't think Mapsui is behaving incorecctly - on the contrary, ArcMap seems to behave in the same way.
